core/server/master/src/main/java/alluxio/master/file/MetadataSyncLock.java

```diff
@@ -0,0 +1,68 @@
+package alluxio.master.file;
+
+import alluxio.AlluxioURI;
+
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class MetadataSyncLock {
```

Comment:
Reasoning on why we need a separated metadata sync lock for metadata sync, instead of using LockedInodePath:
1. LockedInodePath only locks inodes that exist. E.g. doing a metadata sync on /a/b/c/d will only lock / and / -> a, if the directory a doesn't exist. This will result in multiple requests entering the metadata sync.
2.  Today, the thread which handles metadata sync will only do the sync for the metadata sync root. The LockedInodePath will be released after the metadata sync root finishes the sync （https://github.com/Alluxio/alluxio/blob/master/core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java#L401-L402). The children are submitted to a queue and taken care by other worker threads. These worker threads acquire locks separately.  There isn't a single LockedInodePath obtained by the thread throughout the metadata sync process. So it cannot block another metadata sync on the same directory from happening. 
3. Another option is to add a locked path lock in WRITE_EDGE mode for the metadata sync process. This is absolutely a bad solution because it blocks all files system read operations.

Reply:
This MetadataSyncLock locks the path by string instead of name, which addressed 1. But the downside is that such string operations are really inefficient.

Reply:
Yes I think a separate lock table is needed for this, so this solution makes sense. Given that metadata sync operations are already expensive, I'm not too worried about the performance.